### ISSUE TITLE
[Feature request] EasyMotion-like ability to jump to files

### ISSUE THREAD

**Issue by qadzek** [vifm/vifm#1006]
Plugins like [EasyMotion](https://github.com/easymotion/vim-easymotion) allow the user to jump to any location in a Vim buffer very quickly. Similar (Neo)Vim plugins include [Hop](https://github.com/hadronized/hop.nvim), [Sneak](https://github.com/justinmk/vim-sneak) and [Leap](https://github.com/ggandor/leap.nvim).

I'm wondering if there's any interest in a similar feature for Vifm. There was an attempt to add this "Quickjump navigation mode" to [Ranger](https://github.com/ranger/ranger/pull/236).

In my opinion, this feature would make Vifm even better. It would be particularly useful when the cursor is on a file near the top of a long list of visible files, and you want to jump to one of the last visible files.
Current solutions to get there:
- Hold the `j` key for several seconds
- Perform a search: hit `/`, type a couple of characters, and finally press `<CR>`
- Use the relative line numbers, e.g. `23j`

If this feature were implemented, the user would just need to hit a key binding (e.g. `J` or `<C-j>` for files below the cursor). Special letters would appear next to the files (or instead of the relative line numbers). The user would hit the corresponding letter to jump to the desired file. This means they could reach the file in just two keystrokes, possibly without leaving the home row and certainly without reaching for the number keys. 

This screenshot shows how this feature could look:

![vifm_fr](https://github.com/user-attachments/assets/d3d6e2d9-b271-4aab-a849-55d33181cf66)

**Comment by haolian9** [comment]
this is an interesting topic, and i just put some thought on it. instead doing any change to vifm itself, i figured these approaches:
* vifm's lua api + fzf: get the content of a pane, spawn&feed fzf, collect the output of fzf, jump to the picked entry
    * to spawn fzf: 1) we can just switch vifm to background 2) or use tmux's display-pane
* utilize nvim's terminal mode: run a vifm process in nvim's terminal buffer, press some key to enter nvim's terminal mode, so that hop ... motion are available and select an entry, notify vifm the jump and let vifm handle it
    * notify vifm the jump: 1) write some control sequences (well, idk how to name it properly) through the terminal channel 2) or use a dedicated pipe
    * let vifm handle it: 1) idk if vifm read control sequences from stdin natively 2) let lua watch and read the pipe and react to the control sequences.
    * hop ... motion and select an entry: 1) do some work to prune the content on vifm's UI: tab, column, status ... 2) ask vifm the available entries by sending some control sequence to vifm and vifm will send them to stdout so that we can use nvim's termResponse Event to get them 3) ask vifm's lua executor via the pipe and read the pipe on nvim side.

**Comment by xaizek** [comment]
> Plugins like EasyMotion allow the user to jump to any location in a Vim buffer very quickly. Similar (Neo)Vim plugins include Hop, Sneak and Leap.

This reminds me more of labels in Vim-like browsers or PDF viewers, because unlike EasyMotion, labels aren't part of the text itself.  [easybuffer.vim](https://github.com/troydm/easybuffer.vim) has similar navigation.

> in just two keystrokes

I feel like not many Vim users know about <kbd>H</kbd>/<kbd>M</kbd>/<kbd>L</kbd> movement keys which are available in Vifm [as well](https://vifm.info/vimdoc.shtml#vifm-H).

Depending on the file names [<kbd>f</kbd>/<kbd>F</kbd> key](https://vifm.info/vimdoc.shtml#vifm-f) might also work, which is also 2 keys.

***

> vifm's lua api + fzf:

This should work.

> utilize nvim's terminal mode: 

This sounds unreliable, hard to parse something that wasn't intended to be parsed.

***

I think this subject came up before in relation to file views and also last year in https://github.com/vifm/vifm/issues/886 related to menus.  Doesn't seem hard to implement but frankly doesn't look like a pressing issue either, hence I didn't try to do it.  I like the directionality aspect proposed here, by the way.

**Comment by haolian9** [comment]
> This should work.

yeah, i have a proof of concept repo [here](https://github.com/haolian9/cd.vifm), yet it's not perfect as i see delay (1s most time) between selected a entry and jump to the entry.

the `f/F` +`;,` combo is just great and easy to access, i like it.

but i can not deny that i'm intrigued by my second approach, as i often use vifm in nvim, and have my own motion plugin already, it'd be a joyful try.

**Comment by xaizek** [comment]
> something like this? that'd be great.

To that effect.  The picture in the OP is a nice demonstration.

> yeah, i have a proof of concept repo [here](https://github.com/haolian9/cd.vifm), yet it's not perfect as i see delay (1s most time) between selected a entry and jump to the entry.

Does your Vifm include 2d30b394fe3da45cc5684f5048851c36f4620244 commit?  It reduces the delay for invoking Lua callbacks.
